dxp/resource_serving.md

```diff
@@ -0,0 +1,161 @@
+
+#### JavaScript Single File
+
+When a single `.js` file is requested and the `minifierType` is set to `js`, the filter will minify the JavaScript file contents delegating to the [MinifierUtil.minifyJavascript() method](https://github.com/liferay/liferay-portal/blob/master/portal-impl/src/com/liferay/portal/minifier/MinifierUtil.java#L45), which invokes the configured minifier (the OSGi service implementing the [JavaScriptMinifier](https://github.com/liferay/liferay-portal/blob/master/portal-impl/src/com/liferay/portal/minifier/JavaScriptMinifier.java) interface with highest service ranking).
+
+Currently there are two implementations of `JavaScriptMinifier` which come out of the box with DXP: [YahooJavaScriptMinifier](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-minifier/src/main/java/com/frontend/js/minifier/YahooJavaScriptMinifier.java) and [GoogleJavaScriptMinifier](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-minifier/src/main/java/com/frontend/js/minifier/GoogleJavaScriptMinifier.java), being the last one (Google) the default one.
+
+#### CSS Single File
+
+In this case (when a single `.css` file is requested) the filter does three things:
+
+1. Aggregating all imported CSS files (using the `@import url(...);` syntax) into one single resource.
+2. Replacing CSS tokens.
+3. Minifying the resulting CSS content.
+
+The first step simply concatenates the CSS content of the imported files and the host one and updates relative URLs to avoid breaking the resulting CSS.
+
+The second invokes [DynamicCSSUtil.replaceToken()](https://github.com/liferay/liferay-portal/blob/master/portal-impl/src/com/liferay/portal/servlet/filters/dynamiccss/DynamicCSSUtil.java#L47) to substitute the tokens `@base_url@`, `@portal_ctx@` and `@theme_image_path@` by their respective values.
+
+Finally, the last step invokes [MinifierUtil.minifyCss()](https://github.com/liferay/liferay-portal/blob/master/portal-impl/src/com/liferay/portal/minifier/MinifierUtil.java#L37) to minify the generated CSS which, in turn, uses [CSSCompressor](https://github.com/liferay/liferay-portal/blob/master/portal-impl/src/com/liferay/portal/minifier/CSSCompressor.java) (an inlined fork of the old YUI CSS compressor) to minify the served CSS.
+
+### Dynamic CSS Filter
+
+This is a very limited version of the `AggregateFilter` for CSS files. All it does is calling [DynamicCSSUtil.replaceToken()](https://github.com/liferay/liferay-portal/blob/master/portal-impl/src/com/liferay/portal/servlet/filters/dynamiccss/DynamicCSSUtil.java#L47), as in the other filter, to substitute the tokens `@base_url@`, `@portal_ctx@` and `@theme_image_path@` by their respective values.
```

Comment:
"All it does is calling" → "All it does is call".

Line-specific link in this one too, plus backticks.